[Jetpack Content Migration] Migration Events Tracking

salimbraksa opened the pull request:
Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/19701

This PR adds migration flow events tracking. See [Github issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/19701) for a list of migration events tracked.

<img src="https://user-images.githubusercontent.com/9609223/205341289-3f52eff1-4ac8-403b-9ac7-9b6cb1a88b11.png" width=600 />

## Test Instructions

<details><summary><b>Prerequisites: Enable Feature Flags</b></summary>
<p>

**Via the app**
1. Clean install Jetpack and WordPress apps then log into your account
2. My Site > Me > App Settings > Debug
3. Enable "Content Migration" Feature Flag on both apps
4. Enable "Jetpack powered bottom sheet" Feature Flag on WordPress app

**Programmatically**
1. Go to `FeatureFlag.swift` file
2. Enable `contentMigration` and `jetpackPoweredBottomSheet` Feature Flags
3. Build and run both WordPress and Jetpack apps

</p>
</details>

### 1️⃣ Migration Flow Initiation Phase

Build and run Jetpack app. Then, you'll either land on the migration welcome screen or prologue screen:

- **Welcome Screen**: In this case, you should see the following events in the console.

```
🔵 Tracked: migration_content_import_eligibility <eligible: true>
🔵 Tracked: migration_content_import_succeeded
🔵 Tracked: migration_email_triggered
🔵 Tracked: migration_welcome_screen_shown
🔵 Tracked: migration_email_sent
```

- **Jetpack Prologue**: In this case, follow these steps to initiate the migration flow from the WordPress app:
     - Switch to WordPress app
     - Locate any Jetpack powered feature ( e.g Notifications tab )
     - Tap "Jetpack powered" badge
     - Tap "Try the new Jetpack app"
     - **Expect** the migration flow to start on Jetpack app
     - **Expect** to see the following track events in the console:

```
🔵 Tracked: migration_content_import_eligibility <eligible: true>
🔵 Tracked: migration_content_import_failed <error_type: The database couldn't be copied from shared directory>
🔵 Tracked: migration_content_import_succeeded
🔵 Tracked: migration_email_triggered
🔵 Tracked: migration_welcome_screen_shown
🔵 Tracked: migration_email_sent
```
    
### 2️⃣ Welcome Screen Phase

1. Tap "Need help?" button 
2. Dismiss the support screen
3. Tap "Me" button
4. Dismiss the support screen
5. Tap "Continue" button
6. **Expect** the notifications screen to appear
7. **Expect** to see these events in the console

```
🔵 Tracked: migration_welcome_screen_help_button_tapped
🔵 Tracked: migration_welcome_screen_avatar_tapped 
🔵 Tracked: migration_welcome_screen_continue_button_tapped
🔵 Tracked: migration_notifications_screen_shown
```

### 3️⃣ Notifications Screen Phase

There are 3 scenarios in the Notifications screen:

<details><summary><b>Case 1: Continue button tapped and permission granted</b></summary>
<p>

**Expect** to see these console logs:

```
🔵 Tracked: migration_notifications_screen_continue_button_tapped <>
🔵 Tracked: migration_notifications_screen_permission_granted <>
🔵 Tracked: migration_thanks_screen_shown <>
```

</p>
</details>

<details><summary><b>Case 2: Continue button tapped and permission denied</b></summary>
<p>

**Expect** to see these console logs:

```
🔵 Tracked: migration_notifications_screen_continue_button_tapped <>
🔵 Tracked: migration_notifications_screen_permission_denied <>
🔵 Tracked: migration_thanks_screen_shown <>
```

</p>
</details>

<details><summary><b>Case 3: Decide later button tapped</b></summary>

**Expect** to see these console logs:

```
🔵 Tracked: migration_notifications_screen_decide_later_button_tapped <>
🔵 Tracked: migration_thanks_screen_shown <>
```

</details>

### 4️⃣ Thanks Screen Phase

1. Tap "Finish" button
2. **Expect** the main app UI to show up
3. **Expect** to see these console logs

```
🔵 Tracked: migration_thanks_screen_finish_button_tapped <>
🔵 Tracked: migration_please_delete_wordpress_card_shown <>
```

### 5️⃣ Main App Phase

1. Tap "Please delete WordPress" card 
2. **Expect** the "Please delete WordPress" screen to appear
3. Tap "Need help?" button
4. **Expect** the support screen to appear
5. Dismiss the support screen
6. Tap "Got it" button
7. Delete WordPress app
8. Switch back to Jetpack app
9. **Expect** the "Please delete WordPress app" to disappear
10. **Expect** to see these console logs

```
🔵 Tracked: migration_please_delete_wordpress_card_tapped <>
🔵 Tracked: migration_please_delete_wordpress_screen_shown <>
🔵 Tracked: migration_please_delete_wordpress_screen_help_tapped <>
🔵 Tracked: migration_please_delete_wordpress_screen_gotit_tapped <>
🔵 Tracked: migration_please_delete_wordpress_card_hidden <>
```

## Regression Notes
1. Potential unintended areas of impact
None.

11. What I did to test those areas of impact (or what existing automated tests I relied on)
None.

12. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

salimbraksa committed:
Toggle Success Card visibility depending on whether WordPress is installed and whether migration is complete

salimbraksa committed:
Don't show the Success Card if Content Migration flag is disabled

salimbraksa committed:
Use AppConfiguration.isJetpack instead of compiler directive

salimbraksa committed:
Remove track events from getWordPressInstallationState() method

salimbraksa committed:
Add an events tracker for the migration flow

salimbraksa committed:
Refactor the MigrationEmailService to use the new migration events tracker

salimbraksa committed:
Bundle migration events in MigrationEvent enum

salimbraksa committed:
Remove MigratableStateTracker and move its logic to MigrationAnalyticsTracker

salimbraksa committed:
Track migration screen shown event

salimbraksa committed:
Track migration welcome screen button tap events

salimbraksa committed:
Track migration notifications events

salimbraksa committed:
Track thanks screen finish event

salimbraksa committed:
Track migration content import events

salimbraksa committed:
Track please delete wordpress screen events

salimbraksa committed:
Track please delete wordpress card events

salimbraksa committed:
Better error messages when content import fails

salimbraksa committed:
Refactor how migration screen shown event is tracked

salimbraksa committed:
Fix Trailing Whitespace Violation

salimbraksa committed:
Sync with release/21.3 and fix merge conflicts

salimbraksa committed:
Resolve an issue that was causing the "migration_welcome_screen_help_button_tapped" event to be tracked when "Me" avatar is tapped

salimbraksa committed:
Delete migration content export events as they're not needed right now

salimbraksa committed:
Revert changes of few migration files

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Success card/MigrationSuccessCardView.swift:
```diff
@@ -79,11 +79,25 @@ class MigrationSuccessCardView: UIView {
+            Self.cachedShouldShowMigrationSuccessCard = newValue
+        }
+
+        return newValue
```
It's not a good practice to track analytics events inside a read-only property. But it was difficult to find a better place for the events tracking logic considering the time constraints and the fact that the migration feature is temporary.

Also, the `shouldShowMigrationSuccessCard` is called multiple times at runtime. I added `cachedShouldShowMigrationSuccessCard` to ensure the events are fired only when the the value of `shouldShowMigrationSuccessCard` changes.

salimbraksa commented on WordPress/Classes/Utility/Analytics/WPAnalyticsEvent.swift:
```diff
@@ -415,11 +415,6 @@ import Foundation
-    // WordPress to Jetpack Migration
-    case migrationEmailTriggered
-    case migrationEmailSent
-    case migrationEmailFailed
```
Moved these to `MigrationEvent` enum.

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Welcome/MigrationWelcomeViewController.swift:
```diff
@@ -20,13 +22,22 @@ final class MigrationWelcomeViewController: UIViewController {
+            self?.tracker.track(.welcomeScreenHelpButtonTapped)
+            configuration.secondaryHandler?()
+        }
         return actionsView
```
Before these changes, the events tracking logic was executed inside `configuration.primaryHandler` and `configuration.secondaryHandler`.

But since both the "Need help?" and "Me" buttons call `configuration.secondaryHandler?()`, it wasn't possible to differentiate between these 2 events:

```
migration_welcome_screen_help_button_tapped
migration_welcome_screen_avatar_tapped
```

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Welcome/MigrationWelcomeViewController.swift:
```diff
@@ -66,7 +82,11 @@ final class MigrationWelcomeViewController: UIViewController {
+                return
+            }
+            self.tracker.track(.welcomeScreenAvatarTapped)
+            self.viewModel.configuration.actionsConfiguration.secondaryHandler?()
```
As I explained in this [comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/19711/files#r1038388906), the "Me" button also calls `configuration.secondaryHandler?()`

twstokes reviewed (approved):
Great work @salimbraksa! 🚀 

I tested all cases except for Test 3 cases 2 and 3, in case someone else wants to confirm those.

Also, we'll want to consider events [for this new view as well](https://github.com/wordpress-mobile/WordPress-iOS/pull/19710). I can add an issue for this. I don't think it'll be as high priority since we'll be doing [related tracking already](https://github.com/wordpress-mobile/WordPress-iOS/pull/19710/files#diff-bbbf5bdbd178be9be3ed7ba8cea3ae4a19b0b1f28cacdd647d2704014f96b98bR31).

salimbraksa commented:
Thanks for reviewing @twstokes 🙇.

I want to mention that `MigratableStateTracker` was removed and replaced with `MigrationAnalyticsTracker`. The idea is to have one class responsible for tracking all migration events. The method `track` that was in `MigratableStateTracker` was moved to `MigrationAnalyticsTracker` and renamed `trackWordPressMigrationEligibility`. 
[
See `MigrationAnalyticsTracker` implementation.](https://github.com/wordpress-mobile/WordPress-iOS/pull/19711/files#diff-93217cff8e66e084b628c44e69bfdc436cfef776ce627763ea84547da5ae8df0)

twstokes commented:
> Thanks for reviewing @twstokes 🙇.
> 
> I want to mention that `MigratableStateTracker` was removed and replaced with `MigrationAnalyticsTracker`. The idea is to have one class responsible for tracking all migration events. The method `track` that was in `MigratableStateTracker` was moved to `MigrationAnalyticsTracker` and renamed `trackWordPressMigrationEligibility`. [ See `MigrationAnalyticsTracker` implementation.](https://github.com/wordpress-mobile/WordPress-iOS/pull/19711/files#diff-93217cff8e66e084b628c44e69bfdc436cfef776ce627763ea84547da5ae8df0)

Understandable and makes sense to me @salimbraksa - this follows patterns we've done in the past. We'll be sure to adapt #19710 if this PR gets merged first.
